File: src/process/childenv.ts
/**
 * Adapter invariant: a child agent CLI never inherits this process's environment.
 *
 * Ported from spikes/common/childenv.py. Spawning from inside an agent session leaks the
 * parent's session markers, and Claude Code responds to CLAUDE_CODE_CHILD_SESSION by silently
 * disabling transcript persistence -- writing no session file at all. Since the transcript IS
 * the adapter's recovery and audit path, an inherited environment is silent data loss that
 * surfaces much later looking like a parser bug.
 *
 * THAT IS MODE-DEPENDENT, and the mode it holds in is the one this adapter drives (#238).
 * Measured on 2.1.261, twice each:
 *
 *   interactive pty (`entrypoint: cli`)   set -> NO transcript written.  unset -> written.
 *   `claude --print` (`entrypoint: sdk-cli`)  set -> written.  unset -> written.
 *
 * Said explicitly because the easy way to check this claim is `--print`, where it is FALSE --
 * and a reader who checks it that way concludes the guard is obsolete and removes something
 * that is preventing silent data loss in the mode that matters. The turn ran in every case, so
 * what is suppressed is persistence, not the child.
 *
 * The variable is not vestigial: it appears 20 times in the installed 2.1.263 bundle, and
 * `childenvClaims.test.ts` checks that against the binary rather than restating it here. Plain
 * `grep` on the bundle finds nothing -- it is a Mach-O executable -- which is a trap worth
 * knowing for anyone re-checking this by hand.
 *
 * The environment is constructed, never inherited. There is deliberately no default
 * that reaches for process.env behind the caller's back.
 */

export const PARENT_AGENT_PREFIXES = [
  'CLAUDE',
  'CODEX',
  'ANTHROPIC_',
  'OPENAI_',
  'CURSOR_',
  'AIDER_',
  'COPILOT_',
] as const

export const PARENT_AGENT_EXACT = new Set(['AI_AGENT', 'AGENT_SESSION_ID', 'CI'])

/** Not agent markers, but they change how a TUI renders or whether it colours output. */
export const TERMINAL_NOISE = new Set([
  'NO_COLOR',
  'FORCE_COLOR',
  'TERM_PROGRAM',
  'TERM_PROGRAM_VERSION',
  'COLORTERM',
  'TERM_SESSION_ID',
  'ITERM_SESSION_ID',
])

/**
 * Observed leaking from a Claude Code parent on 2026-08-05 (claude 2.1.222). Recorded so
 * a toolchain bump that renames them shows up as a diff rather than a mystery.
 */
export const KNOWN_PARENT_AGENT_VARS = [
  'CLAUDECODE',
  'CLAUDE_CODE_CHILD_SESSION',
  'CLAUDE_CODE_SESSION_ID',
  'CLAUDE_CODE_BRIDGE_SESSION_ID',
  'CLAUDE_CODE_ENTRYPOINT',
  'CLAUDE_CODE_EXECPATH',
  'CLAUDE_CODE_ENABLE_TELEMETRY',
  'CLAUDE_PID',
  'CLAUDE_EFFORT',
  'AI_AGENT',
] as const

/**
 * The ones whose leakage is destructive rather than merely untidy.
 *
 * Read by `assertClean`, so the failure says which it was. Every parent-agent variable is
 * stripped and any leak is a bug, but these two facts are not the same size: an inherited
 * CLAUDE_CODE_ENTRYPOINT is untidy, and an inherited CLAUDE_CODE_CHILD_SESSION means the run
 * that just started is writing no transcript at all -- which is the adapter's recovery and
 * audit path, and which fails much later looking like a parser bug.
 *
 * It was a bare exported constant that nothing read (#238): dead code asserting a fact, which
 * is the shape that rots without anything failing. It is wired now, and the fact is checked
 * against the installed binary in `childenvClaims.test.ts`.
 */
export const TRANSCRIPT_KILLING_VARS = ['CLAUDE_CODE_CHILD_SESSION'] as const

const ALLOWLIST_EXACT = new Set([
  'PATH', 'HOME', 'USER', 'LOGNAME', 'SHELL', 'TMPDIR', 'TZ',
  'LANG', 'LC_ALL', 'LC_CTYPE',
  'SSL_CERT_FILE', 'SSL_CERT_DIR', 'NODE_EXTRA_CA_CERTS',
  'HTTP_PROXY', 'HTTPS_PROXY', 'NO_PROXY',
  'http_proxy', 'https_proxy', 'no_proxy',
  'XDG_CONFIG_HOME', 'XDG_DATA_HOME', 'XDG_CACHE_HOME',
])

const ALLOWLIST_PREFIXES = ['ORCH_'] as const

export type Env = Record<string, string>

export function isParentAgentVar(name: string): boolean {
  return PARENT_AGENT_PREFIXES.some((p) => name.startsWith(p)) || PARENT_AGENT_EXACT.has(name)
}

export interface Discovery {
  present: string[]
  known: string[]
  /** Matches the parent-agent pattern but is not in KNOWN_PARENT_AGENT_VARS. */
  novel: string[]
}

export function discoverParentAgentVars(env: Env = process.env as Env): Discovery {
  const present = Object.keys(env).filter(isParentAgentVar).sort()
  const known = new Set<string>(KNOWN_PARENT_AGENT_VARS)
  return {
    present,
    known: present.filter((n) => known.has(n)),
    novel: present.filter((n) => !known.has(n)),
  }
}

export interface BuildOptions {
  extra?: Env
  rows?: number
  cols?: number
}

function terminalDefaults(rows: number, cols: number): Env {
  return { TERM: 'xterm-256color', COLUMNS: String(cols), LINES: String(rows) }
}

/**
 * Fail loudly rather than let a leak through.
 *
 * If this throws, the child would have run with a parent's markers. For most of them that is
 * an attribution problem; for `TRANSCRIPT_KILLING_VARS` it is silent data loss in the pty mode
 * this adapter drives -- see the header -- so the message says which kind was caught rather
 * than leaving that to be looked up.
 */
function assertClean(env: Env): Env {
  const leaked = Object.keys(env).filter(isParentAgentVar).sort()
  if (leaked.length > 0) {
    const destructive = leaked.filter((n) => (TRANSCRIPT_KILLING_VARS as readonly string[]).includes(n))
    const why = destructive.length
      ? ` — ${destructive.join(', ')} would have left the child writing no transcript at all in pty mode`
      : ''
    throw new Error(`parent-agent variables leaked into child env: ${leaked.join(', ')}${why}`)
  }
  return env
}

/** Denylist. Permissive: keeps proxies, cert bundles and toolchain shims working. */
export function sanitizedCopy(source: Env = process.env as Env, opts: BuildOptions = {}): Env {
  const { extra, rows = 40, cols = 120 } = opts
  const out: Env = {}
  for (const [k, v] of Object.entries(source)) {
    if (v === undefined) continue
    if (isParentAgentVar(k) || TERMINAL_NOISE.has(k)) continue
    out[k] = v
  }
  Object.assign(out, terminalDefaults(rows, cols), extra ?? {})
  return assertClean(out)
}

/** Allowlist. Strict: use when the child must be reproducible rather than merely working. */
export function allowlistEnv(source: Env = process.env as Env, opts: BuildOptions = {}): Env {
  const { extra, rows = 40, cols = 120 } = opts
  const out: Env = {}
  for (const [k, v] of Object.entries(source)) {
    if (v === undefined || isParentAgentVar(k)) continue
    if (ALLOWLIST_EXACT.has(k) || ALLOWLIST_PREFIXES.some((p) => k.startsWith(p))) out[k] = v
  }
  Object.assign(out, terminalDefaults(rows, cols), extra ?? {})
  return assertClean(out)
}

File: spikes/common/childenv.py
"""
Adapter invariant: a child agent CLI never inherits this process's environment.

Rationale (spike 1). Spawning a coding-agent CLI from inside another agent session
leaks the parent's session markers. Claude Code, on seeing CLAUDE_CODE_CHILD_SESSION,
disables transcript persistence entirely and writes no session file. Since the
orchestrator's entire output path is those transcript files, an inherited environment
turns into a silent, far-downstream data loss.

So: the child environment is CONSTRUCTED, never inherited. Two strategies:

  sanitized_copy()  - denylist. Copy the parent env minus anything that identifies a
                      parent agent. Permissive; keeps proxies, cert bundles, toolchain
                      shims working without having to enumerate them.
  allowlist_env()   - allowlist. Only reviewed variables survive. Strict; use when the
                      child must be reproducible rather than merely working.

sanitized_copy() is the default. allowlist_env() is available for when we want to pin
child behaviour down, and the conformance suite should eventually run both.

Nothing here is specific to a transport. When AgentSession lands (step 4) this module
is what its spawn path must call, whichever language the rest ends up in.
"""

import json
import os
import re
import time

# --- the denylist ------------------------------------------------------------------
# Prefixes covering the vendor CLIs and their SDKs. Deliberately broad: a variable we
# strip unnecessarily costs nothing, one we leak can silently disable transcripts.
PARENT_AGENT_PREFIXES = (
    "CLAUDE",
    "CODEX",
    "ANTHROPIC_",
    "OPENAI_",
    "CURSOR_",
    "AIDER_",
    "COPILOT_",
)

# Exact names that don't share a vendor prefix.
PARENT_AGENT_EXACT = frozenset(
    {
        "AI_AGENT",
        "AGENT_SESSION_ID",
        "CI",
    }
)

# Not agent markers, but they change how a TUI renders or whether it colours output.
# Stripped so the child's terminal behaviour depends on the PTY we gave it, not on
# whatever terminal the orchestrator happens to be running under.
TERMINAL_NOISE = frozenset(
    {
        "NO_COLOR",
        "FORCE_COLOR",
        "TERM_PROGRAM",
        "TERM_PROGRAM_VERSION",
        "COLORTERM",
        "TERM_SESSION_ID",
        "ITERM_SESSION_ID",
    }
)

# Confirmed present in a Claude Code session on 2026-08-05 (claude 2.1.222). Recorded
# so a future toolchain bump that renames them shows up as a diff rather than a
# mystery. CLAUDE_CODE_CHILD_SESSION is the one that actually breaks transcripts.
KNOWN_PARENT_AGENT_VARS = (
    "CLAUDECODE",
    "CLAUDE_CODE_CHILD_SESSION",
    "CLAUDE_CODE_SESSION_ID",
    "CLAUDE_CODE_BRIDGE_SESSION_ID",
    "CLAUDE_CODE_ENTRYPOINT",
    "CLAUDE_CODE_EXECPATH",
    "CLAUDE_CODE_ENABLE_TELEMETRY",
    "CLAUDE_PID",
    "CLAUDE_EFFORT",
    "AI_AGENT",
)

# The single variable whose leakage is known to be destructive rather than merely
# untidy. Asserted separately so the regression test can name it.
TRANSCRIPT_KILLING_VARS = ("CLAUDE_CODE_CHILD_SESSION",)

# --- the allowlist -----------------------------------------------------------------
# Reviewed minimum for running a node/rust CLI that has to reach the network, resolve
# a toolchain shim, and read the user's credentials from disk.
ALLOWLIST_EXACT = frozenset(
    {
        "PATH", "HOME", "USER", "LOGNAME", "SHELL", "TMPDIR", "TZ",
        "LANG", "LC_ALL", "LC_CTYPE",
        "SSL_CERT_FILE", "SSL_CERT_DIR", "NODE_EXTRA_CA_CERTS",
        "HTTP_PROXY", "HTTPS_PROXY", "NO_PROXY",
        "http_proxy", "https_proxy", "no_proxy",
        "XDG_CONFIG_HOME", "XDG_DATA_HOME", "XDG_CACHE_HOME",
    }
)
ALLOWLIST_PREFIXES = ("SPIKE_",)


def is_parent_agent_var(name: str) -> bool:
    return name.startswith(PARENT_AGENT_PREFIXES) or name in PARENT_AGENT_EXACT


def discover_parent_agent_vars(env=None):
    """Return parent-agent variables present in `env`, split into ones we already know
    about and ones we don't. Unknown hits are what a toolchain bump looks like."""
    env = os.environ if env is None else env
    present = sorted(n for n in env if is_parent_agent_var(n))
    known = [n for n in present if n in KNOWN_PARENT_AGENT_VARS]
    novel = [n for n in present if n not in KNOWN_PARENT_AGENT_VARS]
    return {"present": present, "known": known, "novel": novel}


def record_discoveries(path, env=None):
    """Append a timestamped discovery record. Values are never written -- names only;
    several of these are session identifiers and one is a telemetry flag."""
    d = discover_parent_agent_vars(env)
    d["recorded_at"] = time.strftime("%Y-%m-%dT%H:%M:%S%z")
    os.makedirs(os.path.dirname(os.path.abspath(path)), exist_ok=True)
    with open(path, "a") as fh:
        fh.write(json.dumps(d, sort_keys=True) + "\n")
    return d


def sanitized_copy(env=None, extra=None, rows=40, cols=120):
    """Parent env minus every parent-agent marker and terminal-identity variable."""
    env = os.environ if env is None else env
    out = {
        k: v
        for k, v in env.items()
        if not is_parent_agent_var(k) and k not in TERMINAL_NOISE
    }
    out.update(_terminal_defaults(rows, cols))
    if extra:
        out.update(extra)
    _assert_clean(out)
    return out


def allowlist_env(env=None, extra=None, rows=40, cols=120):
    """Only reviewed variables survive."""
    env = os.environ if env is None else env
    out = {
        k: v
        for k, v in env.items()
        if (k in ALLOWLIST_EXACT or k.startswith(ALLOWLIST_PREFIXES))
        and not is_parent_agent_var(k)
    }
    out.update(_terminal_defaults(rows, cols))
    if extra:
        out.update(extra)
    _assert_clean(out)
    return out


def _terminal_defaults(rows, cols):
    return {"TERM": "xterm-256color", "COLUMNS": str(cols), "LINES": str(rows)}


def _assert_clean(env):
    """Fail loudly rather than let a leak through. This is the invariant; if it ever
    trips, the child would have run with degraded or absent transcript persistence."""
    leaked = sorted(n for n in env if is_parent_agent_var(n))
    if leaked:
        raise AssertionError(f"parent-agent variables leaked into child env: {leaked}")


if __name__ == "__main__":
    import sys

    d = discover_parent_agent_vars()
    print(json.dumps(d, indent=2, sort_keys=True))
    if d["novel"]:
        print(
            f"\nNOTE: {len(d['novel'])} variable(s) match the parent-agent pattern but are "
            f"not in KNOWN_PARENT_AGENT_VARS. Add them there so the change is recorded.",
            file=sys.stderr,
        )
